### PR TITLE
CPP: Add query for CWE-14 compiler removal of code to clear buffers.

### PR DESCRIPTION
**Pull request by ihsinme**
Good day.

The error in question in this PR is quite common in projects, I tried to minimize the false detection. PR also takes into account the specifics of various compilers.

note that I had to add c.getArgument (2) .toString () = "--force-recompute" to keep the tests running. It was a surprise to me, but the "codeql test run" uses a different line when compiling than the "codeql query run".

I am also worried about your rule.

"Compilation

Compilation of the query and any associated libraries and tests must be resilient to future development of the supported libraries. This means that the functionality cannot use internal libraries, cannot depend on the output of getAQlClass, and cannot make use of regexp matching on toString.
The query and any associated libraries and tests must not cause any compiler warnings to be emitted (such as use of deprecated functionality or missing override annotations). "

if it is a problem then I am ready to weaken this detection.

I want to draw your attention to the following features in the presented PRs.
1. accepted the offer in an optional format.
2. agreed but did not accept, referring to the borrowed code.
https://github.com/lsh123/xmlsec/pull/297  
https://github.com/google/jsonnet/pull/856

### CODE REVIEW

**Comment by jbj** [comment]
Thanks for the submission. We'll find a reviewer.

File: cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.qhelp
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.qhelp:15]: I believe `-fno-builtin-memset` is just _one_ of the flags you can use to prevent the removal of `memset`.
```suggestion
<p>We recommend to use the <code>RtlSecureZeroMemory</code> or <code>memset_s</code> functions, or compilation flags that exclude optimization of <code>memset</code> calls (e.g. -fno-builtin-memset).</p>
```

File: cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.c
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.c:1]:
```suggestion
// BAD: the memset call will probably be removed.
```

File: cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.c
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.c:9]:
```suggestion
// GOOD: in this case the memset will not be removed.
```

File: cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.c
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.c:18]:
```suggestion
// GOOD: in this case the memset will not be removed.
```

File: cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.c
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.c:26]:
```suggestion
// GOOD: in this case the memset will not be removed.
```

File: cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql:36]: Generally, doing control-flow analysis on individual expressions is very expensive. I think what you have here is fine, but there are a couple of improvements that you could make in the future:
- Use [BasicBlock](https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/controlflow/BasicBlocks.qll/type.BasicBlocks$BasicBlock.html)s instead. They also have a [getASuccessor](https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/controlflow/BasicBlocks.qll/predicate.BasicBlocks$BasicBlock$getASuccessor.0.html) predicate that is much faster.
- I get the impression that what you really want here is local dataflow from an allocation to the first argument of this `memset`. That will also be faster than doing expression-level control-flow analysis.

**Review by MathiasVP — COMMENTED** [review]
Thanks for your contribution! This is indeed a problem that is worthy of a CodeQL query :)

I've attached the first round of review comments. If anything is unclear please let me know.

**Comment by ihsinme** [comment]
thanks for your corrections.
I'll try to fix them quickly.

**Comment by intrigus-lgtm** [comment]
@ihsinme a small tip that also helped me:
You can go to the ["Files changed"](https://github.com/github/codeql/pull/4953/files) tab and add multiple suggestions.
This way you can apply multiple suggestions in only one commit :)

File: cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql:35]: Thank you.
However, my idea was to minimize the number of false detections as much as possible. It seemed to me that the flow of control would eliminate complex cases of access to a variable that occur outside the function. (for example, when memory was allocated outside the limits of the current function, and the managed variable received this pointer not from the function argument directly).

so if you agree, I would leave that as it is.

Comment (MathiasVP) [reply cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql:35]: I'm all for minimizing the number of false positives, so if you observe fewer false positives by going with control-flow instead of dataflow that's totally fine by me. 👍

**Comment by intrigus-lgtm** [comment]
All these** commits have only one change.
**Instead of having x commits with only 1 change, you can have only 1 commit with x changes, which looks cleaner.**

To do that, you would go to https://github.com/github/codeql/pull/4953/files and click on "Add suggestion to batch" for every suggestion as seen in step 3.2 [ħere](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request).
Then you can click on "Commit suggestions" to commit them all in only 1 commit :)

Hope this is better to understand :)

**:
https://github.com/github/codeql/pull/4953/commits/4631658e5e4f2bdd26e61e891a11140722bf2a42
https://github.com/github/codeql/pull/4953/commits/76b768f7e0f8e31937665cd92ca21e4eb7af906e
https://github.com/github/codeql/pull/4953/commits/4ba4de3d41eecff5e8d30e9cfd65064bc585869e
https://github.com/github/codeql/pull/4953/commits/0d0ea0c5e15140d97f170b38f227e549487ea7f9

**Comment by ihsinme** [comment]
I cannot see what is the cause of the error.
Can you help me?
QHelp Preview Checks (GitHub Actions) - Internal CI run started by Qlucie

**Comment by criemen** [comment]
The error message is 
```
   [2021-01-14 15:15:31] This is codeql generate query-help -vvv --log-to-stderr --output out --format markdown --search-path . -- cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.qhelp
  [2021-01-14 15:15:31] [WARN] generate query-help> cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql does not have a corresponding qhelp file.
  [2021-01-14 15:15:31] Exception caught at top level: Failed to read path cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql
                        (eventual cause: NoSuchFileException "cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql")
                        com.semmle.util.io.WholeIO.strictread(WholeIO.java:318)
                        com.semmle.prettyprint.doc.QhelpToMD.<init>(QhelpToMD.java:601)
                        com.semmle.prettyprint.doc.QhelpToMD.<init>(QhelpToMD.java:590)
                        com.semmle.cli2.generate.QueryHelpCommand.executeSubcommand(QueryHelpCommand.java:133)
                        com.semmle.cli2.picocli.SubcommandCommon.call(SubcommandCommon.java:430)
                        com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:201)
                        com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:209)
                        com.semmle.cli2.CodeQL.main(CodeQL.java:91)
                        
                         ... caused by:
                        
                        java.nio.file.NoSuchFileException: cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql
                        sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
                        sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
                        sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
                        sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
                        java.nio.file.Files.newByteChannel(Files.java:361)
                        java.nio.file.Files.newByteChannel(Files.java:407)
                        java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
                        java.nio.file.Files.newInputStream(Files.java:152)
                        com.semmle.util.io.WholeIO.read(WholeIO.java:265)
                        com.semmle.util.io.WholeIO.strictread(WholeIO.java:316)
                        com.semmle.prettyprint.doc.QhelpToMD.<init>(QhelpToMD.java:601)
                        com.semmle.prettyprint.doc.QhelpToMD.<init>(QhelpToMD.java:590)
                        com.semmle.cli2.generate.QueryHelpCommand.executeSubcommand(QueryHelpCommand.java:133)
                        com.semmle.cli2.picocli.SubcommandCommon.call(SubcommandCommon.java:430)
                        com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:201)
                        com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:209)
                        com.semmle.cli2.CodeQL.main(CodeQL.java:91)
  A fatal error occurred: Failed to read path cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql
  (eventual cause: NoSuchFileException "cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql")
  Error: Process completed with exit code 2.
```

**Comment by ihsinme** [comment]
Thank you.
the error looks strange to me, I still don't understand what to do in this situation.
it is possible to restart this check.

**Comment by MathiasVP** [comment]
> it is possible to restart this check.

I've restarted it now.

**Comment by ihsinme** [comment]
but unfortunately the error has not gone away.
if the error message remains the same, then I can assume there is a path problem in the test itself.
for my part I revisited the ql qhelp file paths again and found no problem.
I would be grateful if you could suggest possible actions on my part, or provide additional information to reproduce the problem on my side.

**Comment by MathiasVP** [comment]
> for my part I revisited the ql qhelp file paths again and found no problem.

Thanks for looking at the problem. I'll take a look at it later today.

**Comment by ihsinme** [comment]
sorry Sorry sorry ....

**Comment by MathiasVP** [comment]
> sorry Sorry sorry ....

No worries! We've identified a problem with the PR check, so it's totally not your fault :) We'll have a fix on Monday and then we can merge your query!

**Comment by MathiasVP** [comment]
The failing PR check is still being fixed. I see you've opened https://github.com/github/securitylab/issues/237 to get a bounty for getting the query merged, and the ambition level there is up to you.
You can create follow-up PRs to make the query more production-ready, following the steps in https://github.com/github/codeql/blob/main/docs/supported-queries.md.

Until then, we'll continue to review any PRs you make to improve the query. It would be very helpful if you could explicitly state in https://github.com/github/securitylab/issues/237 once you're done improving the query.

**Comment by ihsinme** [comment]
good day.
I'm not experienced enough yet so sorry for the simple questions.
unless I currently plan on improving this request. Did I understand correctly that I need to write in the request for reward that I have finished improving the request?

**Comment by MathiasVP** [comment]
> good day.
> 
> I'm not experienced enough yet so sorry for the simple questions.
> 
> unless I currently plan on improving this request. Did I understand correctly that I need to write in the request for reward that I have finished improving the request?

Yes. If you don't plan on doing further improvements to the query please say so in https://github.com/github/securitylab/issues/237

**Comment by kevinbackhouse** [comment]
This query has a lot of false positives. [Click here for a typical example](https://lgtm.com/projects/g/numpy/numpy/snapshot/024bc72dbaf5027c861ad402ad352cdcd8efcc0e/files/numpy/core/src/common/numpyos.c?sort=name&dir=ASC&mode=heatmap#x1c45f8b48fb6bf0d:1). That `memset` won't get removed by the compiler because it isn't erasing a stack buffer that's about to go out of scope. I would suggest trying the `StackAddress` library.

I have seen [one good result](https://lgtm.com/projects/g/bitcoin/bitcoin/snapshot/07fa68c6db80e59aeffe39c823b0f6c00bcca1ee/files/src/secp256k1/src/ecmult_gen_impl.h?sort=name&dir=ASC&mode=heatmap#x80cb954d9dda84a2:1). But I am puzzled why the query doesn't also tag the `memset(nonce32, 0, 32)` a few line down, which looks like an identical pattern to me.

**Comment by ihsinme** [comment]
> This query has a lot of false positives. [Click here for a typical example](https://lgtm.com/projects/g/numpy/numpy/snapshot/024bc72dbaf5027c861ad402ad352cdcd8efcc0e/files/numpy/core/src/common/numpyos.c?sort=name&dir=ASC&mode=heatmap#x1c45f8b48fb6bf0d:1). That `memset` won't get removed by the compiler because it isn't erasing a stack buffer that's about to go out of scope. I would suggest trying the `StackAddress` library.
> 

indeed this is a false detection. it probably comes from a rather long chain of heap bindings. when two or more variables recursively assign each other a heap value. I'll try to fix it using `StackAddress`. Thank you.

> I have seen [one good result](https://lgtm.com/projects/g/bitcoin/bitcoin/snapshot/07fa68c6db80e59aeffe39c823b0f6c00bcca1ee/files/src/secp256k1/src/ecmult_gen_impl.h?sort=name&dir=ASC&mode=heatmap#x80cb954d9dda84a2:1). But I am puzzled why the query doesn't also tag the `memset(nonce32, 0, 32)` a few line down, which looks like an identical pattern to me.

this place `memset (nonce32, 0, 32)` is skipped, probably because the initialization takes place inside another function, I don't know yet whether it is worth adding this detection. in my opinion it will please other false positives. but you definitely know better.

**Comment by ihsinme** [comment]
I corrected, I expect your reaction.
I think that `StackAddress` may in the future eliminate the need to use some of my predicates in this query, but I'm not ready for that yet.

I consider myself a newbie in your environment, so thanks for the fixes, they add experience.

**Comment by ihsinme** [comment]
good day.
maybe I have something else to do so that the check is launched and the PR is accepted?

**Comment by MathiasVP** [comment]
We are still evaluating the query. https://github.com/github/securitylab/issues/237 shows the current state of the evaluation process.

**Comment by MathiasVP** [comment]
> this place `memset (nonce32, 0, 32)` is skipped, probably because the initialization takes place inside another function, I don't know yet whether it is worth adding this detection. in my opinion it will please other false positives. but you definitely know better.

If you want to handle this case, you can do so by changing your characteristic predicate from this:

```codeql
CompilerRemovaMemset() {
  this.getTarget().hasGlobalOrStdName("memset") and
  exists(DataFlow::Node source, DataFlow::Node sink, LocalVariable isv, Expr exp |
    DataFlow::localFlow(source, sink) and
    this.getArgument(0) = isv.getAnAccess() and
    source.asExpr() = exp and
    exp.getLocation().getEndLine() < this.getArgument(0).getLocation().getStartLine() and
    sink.asExpr() = this.getArgument(0)
  )
```

to this:
```codeql
CompilerRemovaMemset() {
  this.getTarget().hasGlobalOrStdName("memset") and
  exists(DataFlow::Node source, DataFlow::Node sink, LocalVariable isv, Expr exp |
    DataFlow::localFlow(source, sink) and
    this.getArgument(0) = isv.getAnAccess() and
    (
      source.asExpr() = exp
      or
      // handle the case where exp is defined an address being passed into some function.
      source.asDefiningArgument() = exp
    ) and
    exp.getLocation().getEndLine() < this.getArgument(0).getLocation().getStartLine() and
    sink.asExpr() = this.getArgument(0)
  )
}
```

I'll leave it up to you whether this change is worth it or not.

**Comment by ihsinme** [comment]
thanks for the help.
I'll take a look.

**Comment by ihsinme** [comment]
it really gives additional error detection.
I do not yet see the possibility of an additional false detection.
so once again I thank you for your help, I hope in future PR I will need it less.
and I accept this change.

I definitely need to read more about `Expr`. (

**Review by MathiasVP — APPROVED** [review]

**Comment by kevinbackhouse** [comment]
@ihsinme: just by coincidence I am currently writing some code that handles passwords, so I tried using `memset_s` to clear the buffer. But, as far as I can tell, `memset_s` isn't defined on Linux. Do you know what the recommended way to do this is on Linux? (This question isn't related to the bounty process.)

**Comment by ihsinme** [comment]
> @ihsinme: just by coincidence I am currently writing some code that handles passwords, so I tried using `memset_s` to clear the buffer. But, as far as I can tell, `memset_s` isn't defined on Linux. Do you know what the recommended way to do this is on Linux? (This question isn't related to the bounty process.)

it is optimal to use compile flag `-fno-builtin-memset`. which will exclude deletion of `memset`.

a more hardcore method is organizing access to array elements after zeroing.

**Comment by jbj** [comment]
There's [a good USENIX paper on how to use `memset` securely](https://www.usenix.org/system/files/conference/usenixsecurity17/sec17-yang.pdf). It's definitely worth linking to from the qhelp if/when we make a non-experimental version of this query.

**Pull request merged**
